_ext/eclipse-base/src/main/java/com/diffplug/spotless/extra/eclipse/base/service/SingleSlf4JService.java

```diff
@@ -0,0 +1,532 @@
+				create(() -> delegate.isInfoEnabled(), m -> delegate.info(m), (m, e) -> delegate.info(m, e)));
+		logLevel2methods.put(LogLevel.TRACE,
+				create(() -> delegate.isTraceEnabled(), m -> delegate.trace(m), (m, e) -> delegate.trace(m, e)));
+		logLevel2methods.put(LogLevel.WARN,
+				create(() -> delegate.isWarnEnabled(), m -> delegate.warn(m), (m, e) -> delegate.warn(m, e)));
+		listener = new HashSet<LogListener>();
+		this.messageCustomizer = messageCustomizer;
+	}
+
+	@Override
+	@Deprecated
+	//Backward compatibility with Eclipse OSGI 3.12
+	public void log(int level, String message) {
+		log(this, level, message);
+	}
+
+	@Override
+	@Deprecated
+	//Backward compatibility with Eclipse OSGI 3.12
+	public void log(int level, String message, @Nullable Throwable exception) {
+		log(this, level, message, exception);
+	}
+
+	@SuppressWarnings("rawtypes")
+	@Override
+	@Deprecated
+	//Backward compatibility with Eclipse OSGI 3.12
+	public void log(ServiceReference sr, int level, String message) {
+		log(this, level, message);
+	}
+
+	@SuppressWarnings("rawtypes")
+	@Override
+	@Deprecated
+	//Backward compatibility with Eclipse OSGI 3.12
+	public void log(ServiceReference sr, int level, String message, Throwable exception) {
+		log(this, level, message, exception);
+	}
+
+	@Override
+	public void log(Object context, int level, String message) {
+		log(context, level, message, null);
+	}
+
+	@Override
+	public void log(Object context, int level, String message, @Nullable Throwable exception) {
+		LogLevel logLevel = convertDeprectatedOsgiLevel(level);
+		log(new SimpleLogEntry(logLevel, message, exception));
+	}
+
+	@Override
+	public boolean isLoggable(int level) {
+		LogLevel logLevel = convertDeprectatedOsgiLevel(level);
+		return logLevel2methods.get(logLevel).isEnabled();
+	}
+
+	@SuppressWarnings("deprecation") ////Backward compatibility with Eclipse OSGI 3.12
+	private static LogLevel convertDeprectatedOsgiLevel(int level) {
+		switch (level) {
+		case SingleSlf4JService.LOG_DEBUG:
+			return LogLevel.DEBUG;
+		case SingleSlf4JService.LOG_INFO:
+			return LogLevel.INFO;
+		case SingleSlf4JService.LOG_ERROR:
+			return LogLevel.ERROR;
+		case SingleSlf4JService.LOG_WARNING:
+			return LogLevel.WARN;
+		default:
+			return LogLevel.AUDIT;
+		}
+	}
+
+	@Override
+	public String getName() {
+		return delegate.getName();
+	}
+
+	@Override
+	public Logger getLogger(String loggerName) {
+		return this;
+	}
+
+	@Override
+	public Logger getLogger(Bundle bundle, String loggerName) {
+		return this;
+	}
+
+	@Override
+	public void addLogListener(LogListener listener) {
+		synchronized (this.listener) {
+			this.listener.add(listener);
+		}
+	}
+
+	@Override
+	public void removeLogListener(LogListener listener) {
+		synchronized (this.listener) {
+			this.listener.remove(listener);
+		}
+	}
+
+	public void log(LogEntry entry) {
+		synchronized (listener) {
+			listener.stream().forEach(l -> l.logged(entry));
+		}
```

Comment:
Consider using a [`ReadWriteLock`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReadWriteLock.html) instead of synchronizing on `listener`.
If you use a [`ReadWriteLock`] you can have multiple loggers logging at the same time multithreaded.

Reply:
Principally a good idea, but I think Java 8 does not provide guards for it, or do they?
If not, I find to verbose (guards implementation) or to error prone (without guards).
Eclipse code normally always check whether a debug is enabled. The formatter anyhow just log warnings. And most of the underlying Eclipse functionality is stubbed by my framework. So this should really not be performance critical.
OK?

Reply:
Could it be a bottleneck in a large multi-project build where spotless is running in parallel on a lot of projects?